sgx-types/src/secs.rs

```diff
@@ -39,34 +39,47 @@ impl Spec {
     ///
     /// # Safety
     /// This function is unsafe because it does not check if the `CPUID` instruction
-    /// is available before issuing it.
-    pub unsafe fn max_enc_size() -> Option<NonZeroU64> {
+    /// is available before issuing it. According to common practice, these calls
+    /// are wrapped in an unsafe block without making the whole function unsafe.
+    pub fn max_enc_size() -> Option<NonZeroU64> {
         const LEAF_MAX_PARAM: u32 = 0x0;
         const LEAF_SGX_SUPPORT: u32 = 0x07;
         const SUBLEAF_SGX_SUPPORT: u32 = 0x0;
         const LEAF_MAX_ENCL_SIZE: u32 = 0x12;
         const SUBLEAF_MAX_ENCL_SIZE: u32 = 0x0;
 
         // Test for max leaf size
-        let res = core::arch::x86_64::__get_cpuid_max(LEAF_MAX_PARAM);
+        let res = unsafe { core::arch::x86_64::__get_cpuid_max(LEAF_MAX_PARAM) };
         let max_leaf = res.0;
         if max_leaf < LEAF_SGX_SUPPORT || max_leaf < LEAF_MAX_ENCL_SIZE {
             return None;
         }
 
         // Test for SGX support
-        let res = core::arch::x86_64::__cpuid_count(LEAF_SGX_SUPPORT, SUBLEAF_SGX_SUPPORT);
+        let res = unsafe { core::arch::x86_64::__cpuid_count(LEAF_SGX_SUPPORT, SUBLEAF_SGX_SUPPORT) };
         if res.ebx & (1 << 2) == 0 {
             return None;
         }
 
         // Test for max enclave size
-        let res = core::arch::x86_64::__cpuid_count(LEAF_MAX_ENCL_SIZE, SUBLEAF_MAX_ENCL_SIZE);
+        let res = unsafe { core::arch::x86_64::__cpuid_count(LEAF_MAX_ENCL_SIZE, SUBLEAF_MAX_ENCL_SIZE) };
         let max_size: u64 = 1 << (res.edx >> 8 as u8) as u64;
         Some(NonZeroU64::new(max_size).unwrap())
     }
 }
 
+impl Default for Spec {
+    /// # Safety
+    /// This function involves calling the `CPUID` instruction, which is unsafe because it
+    /// does not check whether this instruction is available before issuing it.
```

Comment:
This safety documentation needs to go on `max_enc_size()`. It should explain that although issuing `CPUID` will crash on unsupported CPUs, this is extremely rare and common practice is to just wrap `CPUID` in unsafe blocks and ignore the unsafety. You can also link to this issue for more details: https://github.com/rust-lang/rust/issues/60123

Reply:
I had it on `max_enc_size()` also, but I've now expanded the documentation on `max_enc_size()` and now removed it completely from `Spec::default()`.

Reply:
Please use this text:

```
This function is technically unsafe because it does not check if the `CPUID` instruction is available before issuing it. This could result in a crash on some very old CPUs. However, the only modern context where it could crash is environments like SGX or some virtualized CPUs. But even in these contexts it is common to trap and emulate the instruction.

Therefore, it is common practice to ignore the test to see if the `CPUID` instruction is available and just issue it anyway. For this reason, we are marking this function as safe. For more background to the state of `CPUID` in Rust, see: https://github.com/rust-lang/rust/issues/60123
```

Also, please mark `max_enc_size()` as unsafe in a separate commit. The commit message should contain the above text as well.